relayer/application_relayer.go

```diff
@@ -102,30 +106,84 @@ func newApplicationRelayer(
 	checkpointManager := checkpoint.NewCheckpointManager(logger, db, sub, relayerID, startingHeight)
 	checkpointManager.Run()
 
-	ar := applicationRelayer{
+	ar := ApplicationRelayer{
 		logger:            logger,
 		metrics:           metrics,
 		network:           network,
 		messageCreator:    messageCreator,
-		responseChan:      responseChan,
 		sourceBlockchain:  sourceBlockchain,
 		relayerID:         relayerID,
 		signingSubnetID:   signingSubnet,
 		warpQuorum:        quorum,
 		checkpointManager: checkpointManager,
+		currentRequestID:  rand.Uint32(), // TODONOW: pass via ctor
+		lock:              &sync.RWMutex{},
 	}
 
 	return &ar, nil
 }
 
-func (r *applicationRelayer) relayMessage(
-	unsignedMessage *avalancheWarp.UnsignedMessage,
+// Process [msgs] at height [height] by relaying each message to the destination chain.
+// Checkpoints the height with the checkpoint manager when all messages are relayed.
+func (r *ApplicationRelayer) ProcessHeight(height uint64, msgs []messages.MessageHandler) error {
+	var eg errgroup.Group
+	for _, msg := range msgs {
+		eg.Go(func() error {
+			return r.ProcessMessage(msg)
+		})
+	}
+	if err := eg.Wait(); err != nil {
+		r.logger.Error(
+			"Failed to process block",
+			zap.Uint64("height", height),
+			zap.String("relayerID", r.relayerID.ID.String()),
+			zap.Error(err),
+		)
+		return err
+	}
+	r.checkpointManager.StageCommittedHeight(height)
+	r.logger.Debug(
+		"Processed block",
+		zap.Uint64("height", height),
+		zap.String("sourceBlockchainID", r.relayerID.SourceBlockchainID.String()),
+		zap.String("relayerID", r.relayerID.ID.String()),
+		zap.Int("numMessages", len(msgs)),
+	)
+	return nil
+}
+
+// Relays a message to the destination chain. Does not checkpoint the height.
+func (r *ApplicationRelayer) ProcessMessage(msg messages.MessageHandler) error {
+	// Increment the request ID. Make sure we don't hold the lock while we relay the message.
+	r.lock.Lock()
+	r.currentRequestID++
+	r.lock.Unlock()
+
+	err := r.relayMessage(
+		r.currentRequestID,
```

Comment:
Can more than one goroutine call ProcessMessage (on different messages) at the same time? If yes, then this could result in more than one relayMessage calls to use the same request ID. 

If the request ID needs to be distinct on each call, then the currentRequestID should be copied to a local variable inside the critical section above in between the lock/unlock. 

I also assume that the ordering between calls to relayMessage calls doesn't matter (or is provided in some other parts of the system).

Reply:
Good catch. Yes, requestID needs to be distinct on each call to `ProcessMessage`, but the ordering does not matter. Assigning to a local var in the critical section as suggested addressed this.